designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,312 @@
+kind: NodePool
+metadata:
+  name: default
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    budgets:
+    - schedule: "* * * * *"
+      reason: drifted 
+      nodes: 10
+    # For all other reasons , only allow 5 nodes to be disrupted at a time
+    - nodes: 5
+      schedule: "* * * * *"
+
+```
+
+
+#### Pros and Cons
+👍 Simplicity and Clarity: Offers a straightforward and clear mapping of budget to disruption reason.
+👎 Increased Configuration Overhead: Requires duplicating settings for multiple reasons, increasing setup complexity.
+👎 Less Flexibility: Lacks the flexibility to share a budget across multiple reasons.
+
+#### Pros and Cons for List and Per Reason Definitions
+Some of the Pros and Cons are shared by both list and single reason, as they have the same advantages and disadvantages in comparison to Per Reason Controls 
+* 👍👍 Extends Existing API:  No Breaking API Changes, completely backwards compatible
+* 👍 No Nesting Required: Leaves budgets at the top level of the api.
+* 👎 Limited Generalization of Reason Controls: With reason being clearly tied to budgets, and other api logic being driven by disruption reason, we lose the chance to generalize per Reason controls. If we ever decide we need a place per action,  there will be some duplication for reason. 
+
+
+### Preferred Option: List Approach
+Given the comparison, the preferred design is the List Approach. It provides the necessary flexibility for managing multiple disruption reasons under a single budget, while reducing configuration complexity. This approach extends the existing API without introducing breaking changes and simplifies management for scenarios where multiple disruption reasons share similar constraints.
+
+### Approach: Defining Per Reason Controls  
+Ideally, we could move all generic controls that easily map into other reasons into one set of reason controls, this applies to budgets and other various disruption controls that could be more generic. 
+#### Example 
+
+```yaml 
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: example-nodepool
+spec:
+  disruption:
+    defaults:
+      budgets: 
+        - nodes: 10% 
+          schedule: "0 0 1 * *"
+          duration: 1h 
+    consolidation:
+      consolidationPolicy: WhenUnderutilized
+      disruptAfter: "30m"
+    drift:
+      budgets:
+        - nodes: "20%"
+          schedule: "0 0 1 * *"
+          duration: "1h"
+        - nodes: "10%"
+          schedule: "0 0 * * 0"
+          duration: "2h"
+        - nodes: "50%" 
+          schedule: "@monthly"
+```
+#### Considerations 
+Some of the API choices for a given reason seem to follow a similar pattern. These include ConsolidateAfter, ExpireAfter, and there are discussions about introducing a global DisruptAfter. Moreover, when discussing disruption budgets, we talk about adding behavior for each reason. It appears there is a need for disruption controls within the budgets for each reason, not just overall.
```

Comment:
May not make as much sense if we move the expireAfter into the NodeClaimTemplate itself, but considering this isn't the recommended option, I think we're good here.

Reply:
Agreed.